consensus/hotstuff/validator/validator.go

```diff
@@ -12,74 +12,147 @@ import (
 
 // Validator is responsible for validating QC, Block and Vote
 type Validator struct {
-	// TODO: change to hotstuff.Replicas when front-loading QC verification
-	// https://github.com/onflow/flow-go/pull/2328#discussion_r866494368
-	committee hotstuff.DynamicCommittee
-	forks     hotstuff.ForksReader
+	committee hotstuff.Replicas
 	verifier  hotstuff.Verifier
 }
 
 var _ hotstuff.Validator = (*Validator)(nil)
 
 // New creates a new Validator instance
 func New(
-	committee hotstuff.DynamicCommittee,
-	forks hotstuff.ForksReader,
+	committee hotstuff.Replicas,
 	verifier hotstuff.Verifier,
 ) *Validator {
 	return &Validator{
 		committee: committee,
-		forks:     forks,
 		verifier:  verifier,
 	}
 }
 
-// ValidateQC checks the validity of a QC for a given block. Inputs:
-//  * qc - the qc to be validated
-//  * block - the block that the qc is pointing to
+// ValidateTC validates the TimeoutCertificate `tc`.
 // During normal operations, the following error returns are expected:
-//  * model.InvalidBlockError if the QC is invalid
-func (v *Validator) ValidateQC(qc *flow.QuorumCertificate, block *model.Block) error {
-	if qc.BlockID != block.BlockID {
-		// Sanity check! Failing indicates a bug in the higher-level logic
-		return fmt.Errorf("qc.BlockID %s doesn't match block's ID %s", qc.BlockID, block.BlockID)
+//  * model.InvalidTCError if the TC is invalid
+//  * model.ErrViewForUnknownEpoch if the TC refers unknown epoch
+// Any other error should be treated as exception
+func (v *Validator) ValidateTC(tc *flow.TimeoutCertificate) error {
+	newestQC := tc.NewestQC
+
+	// The TC's view cannot be smaller than the view of the QC it contains.
+	// Note: we specifically allow for the TC to have the same view as the highest QC.
+	// This is useful as a fallback, because it allows replicas other than the designated
+	// leader to also collect votes and generate a QC.
```

Comment:
This explanation doesn't make sense to me.

We're saying that view `n` timed out, but the leader `L` for view `n+1` was able to collect votes for view `n`, so that `L` has:
* `QC_n` with view `n`
* `TC_n` with view `n`

If this happens, `L` may embed the `QC_n` in `TC_n`, in order to make progress using `QC_n` more quickly. But, why would `L` do this, when `L` could instead use `QC_n` directly and omit the TC altogether?

Reply:
Additionally, looking at [`ValidateProposal`](https://github.com/onflow/flow-go/pull/2456/files#diff-8e933e078b2e6031be4617d76cfe25046e38ac012720b757b382869e6a509aeaR218-R239), the case described here would be invalid and result in an error.

Suppose we have a `TC_n` containing a highest QC `QC_n`:

1. If we embed `QC_n` in the proposal, then `lastViewSuccessful` is true on line 218 and we return an error on line 238 because an unnecessary TC is included (check on line 236).

2. If we don't embed `QC_n` in the proposal, we must include a QC with view lower than `n` and we return an error line 234 (check on line 233).

Reply:
Hm, I am not 100% sure here. This comment was suggested by Alex and I think it's relevant for future extensions of the protocol where we might allow fallback leaders. 

@AlexHentschel

Reply:
Yes, this ultimately for a performance optimization. While I don't necessarily want to implement the optimization right away, it would be nice to already implement our code such that it is compatible with this optimization. Up to now, my proposal wasn't written up, so I understand that there is a lot of confusion around it. Spent some time today for properly documenting it: https://www.notion.so/dapperlabs/Optimization-Reducing-impact-of-offline-primaries-ffa994aafcdd43ad99652e9c3b47a951

Regarding your questions @jordanschalm:
* > why would L do this, when L could instead use QC_n directly and omit the TC altogether

   I hope the notion doc clarifies this. 
* > case described here would be invalid and result in an error

  I think it is important to differentiate between the QC included in a block and the QC included in a TC: ![Multiple Vote Aggregators Optimization 3](https://user-images.githubusercontent.com/26322303/172505341-d5e30edf-6841-4014-829d-3aaa4ce95ab3.png)
  * We have a requirement that if `QC_1` is for the prior view, there should not be including a TC.  
  * Lets look at the situation you are describing: 
     * `TC_n` containing a highest `QC_n`. I don't think we have specified anywhere that the proposer for view `N+1` must use the TC. If it has a QC for the last round (here `QC_n`) it can just use that and omit the TC. This is actually preferable anyway, because a TC is probably more than twice as expensive to verify than a QC. 
     * The primary needs some logic to prevent it from accidentally including the TC as well as the QC both for the previous view. Though, I think this logic is very simple (only a few lines of code). Furthermore, I would almost guess that we would add this logic anyway even without the optimization to improve modularity of the code.

Reply:
Thank you for the detailed explanation - I understand now. The piece I was missing was that the next proposal is two views later, and the diagrams in the notion doc make the flow very clear 👌